src/app/core/services/guided-tour/guided-tour-step.service.ts

```diff
@@ -0,0 +1,329 @@
+      .resolveComponentFactory(GuidedTourStepComponent)
+      .create(this._injector);
+
+    this._appRef.attachView(ref.hostView);
+    const domElem = (ref.hostView as EmbeddedViewRef<any>).rootNodes[0] as HTMLElement;
+    document.body.appendChild(domElem);
+
+    const instance: GuidedTourStepComponent = ref.instance;
+    instance.step = step;
+    ref.changeDetectorRef.detectChanges();
+    step.stepInstance = instance;
+
+    this.refMap[step.name] = ref;
+  }
+
+  private _subscribeToStepsUpdates(): void {
+    this._guidedTourStepsContainerService.stepHasBeenModified.subscribe(updatedStep => {
+      if (this._currentStep && this._currentStep.name === updatedStep.name) {
+        this._currentStep = updatedStep;
+      }
+    });
+  }
+
+  private _tryShowStep(actionType: StepActionType): void {
+    const stepRoute = this._guidedTourStepsContainerService.getStepRoute(actionType);
+    if (stepRoute) {
+      this._router.navigate([stepRoute]);
+    }
+
+    const timeout = this._guidedTourOptionsService.getWaitingTime();
+    setTimeout(() => {
+      try {
+        this._showStep(actionType);
+      } catch (error) {
+        this.close();
+      }
+    }, timeout);
+  }
+
+  private _showStep(actionType: StepActionType): void {
+    this._currentStep = this._guidedTourStepsContainerService.getStep(actionType);
+
+    // Scroll the element to get it visible if it's in a scrollable element
+    this._scrollIfElementBeyondOtherElements();
+    this._drawStep(this._currentStep);
+    this._scrollWhenTargetOrStepAreHiddenBottom();
+    this._scrollWhenTargetOrStepAreHiddenTop();
+    this._notifyStepClicked(actionType);
+  }
+
+  private _notifyStepClicked(actionType: StepActionType): void {
+    const stepInfo: GuidedTourStepInfo = {
+      number: this._guidedTourStepsContainerService.getStepNumber(this._currentStep.name),
+      name: this._currentStep.name,
+      route: this._currentStep.route,
+      actionType,
+    };
+    this._stepsObserver.next(stepInfo);
+  }
+
+  private _notifyTourIsFinished(): void {
+    if (this._currentStep) {
+      this._currentStep.tourDone.emit();
+    }
+    this._stepsObserver.complete();
+  }
+
+  private _removeCurrentStep(): void {
+    if (this._currentStep) {
+      this._appRef.detachView(this.refMap[this._currentStep.name].hostView);
+      this.refMap[this._currentStep.name].destroy();
+    }
+  }
+
+  private _scrollWhenTargetOrStepAreHiddenBottom(): void {
+    const totalTargetBottom = this._getMaxTargetAndStepBottomPosition();
+    if (totalTargetBottom > this._winBottomPosition) {
+      window.scrollBy(0, totalTargetBottom - this._winBottomPosition);
+    }
+  }
+
+  private _scrollWhenTargetOrStepAreHiddenTop(): void {
+    const totalTargetTop = this._getMaxTargetAndStepTopPosition();
+    if (totalTargetTop < this._winTopPosition) {
+      window.scrollBy(0, totalTargetTop - this._winTopPosition);
+    }
+  }
+
+  private _getMaxTargetAndStepBottomPosition(): number {
```

Comment:
It looks like something that could be done from stylesheets. Can you explain this?

Reply:
We should rely on styles wherever possible. Trying to manipulate DOM and CSS from the code should be the last thing we should try if everything else fails.